### ISSUE TITLE
A few questions about the implementation.

### ISSUE THREAD

**Issue by fuyw** [eth-ait/sfp#1]
Hi,

Thanks for the nice work. I want to ask some questions about the implementation details:

(1) Why do we use `next_observation` instead of `observation` when we compute the lambda loss, and why do we use the parameter `epsilon` and `beta` here:

https://github.com/eth-ait/sfp/blob/b8eb53b24331e7379c5e24a27a6f4b99f445e2fb/agent.py#L273

(2) When we compute the critic loss:

https://github.com/eth-ait/sfp/blob/b8eb53b24331e7379c5e24a27a6f4b99f445e2fb/agent.py#L203

can we simply compute a weight target Q function:

```python
lambda_ = self.ac.lambda_net(o2)
a2_prior = self.temporal_prior.sample(o2, last_action=la)

q1_pi_targ = self.ac_targ.q1(o2, a2)
q2_pi_targ = self.ac_targ.q2(o2, a2)
q_pi_target = torch.min(q1_pi_targ, q2_pi_targ)

prior_q1_target = self.ac_targ.q1(o2, a2_prior)
prior_q2_target = self.ac_targ.q2(o2, a2_prior)
prior_q_target = torch.min(prior_q1_target, prior_q2_target)

q_target = lambda_* prior_q_target + (1 - lambda_) * q_pi_target
```

**Comment by marbaga** [comment]
Hi, thank you for your interest!

(1) This is mainly an implementation detail. The prior is conditioned on past actions by default. As each training sample traditionally includes $(s_t, a_t, s_{t+1})$ tuples, we compute the prior loss on $(a_t, s_{t+1})$. It would also be possible to compute it on $(a_{t-1}, s_t)$, but that would require modifying the buffer sampling procedure to include past actions in training samples. The resulting behavior should be equivalent. $\epsilon$ is a scaling parameter for the learning rate of the mixing network, which is substantially lower than that of actor/critic networks for stability (see Appendix K.4 in the paper). $\beta$ is a regularizing hyperparameter that would encourage high outputs from the mixing network, but in default configurations it is set to zero, and has therefore no effect.

(2) Thanks for the great question. In expectation, I believe that the two implementations should be equivalent, although we only experimented with the current one. Some side-effects might be introduced by removing the sampling step, but estimating the mixture directly could also potentially help. In case you evaluated this empirically, we would be happy to learn about the results.

**Comment by fuyw** [comment]
Many thanks for the reply.
